pkgs/dart_model/benchmark/json_buffer.dart

```diff
@@ -39,3 +175,191 @@ class JsonBufferDeserializeBenchmark extends BenchmarkBase {
     deserialized = JsonBuffer.deserialize(serialized!);
   }
 }
+
+/// An interface.
+extension type Interface.fromJson(Map<String, Object?> node) {
+  Interface.slow({
+    Map<String, Member>? members,
+    Properties? properties,
+  }) : this.fromJson({
+          if (members != null) 'members': members,
+          if (properties != null) 'properties': properties,
+        });
+
+  Interface({
+    Map<String, Member>? members,
+    Properties? properties,
+  }) : this.fromJson(LazyMap(
+            [
+              if (members != null) 'members',
+              if (properties != null) 'properties',
+            ],
+            (key) => switch (key) {
+                  'members' => members,
+                  'properties' => properties,
+                  _ => null,
+                }));
+
+  /// Map of members by name.
+  Map<String, Member> get members => (node['members'] as Map).cast();
+
+  /// The properties of this interface.
+  Properties get properties => node['properties'] as Properties;
+}
+
+extension type InterfaceBuilder._(JsonBuffer buffer) {
+  factory InterfaceBuilder() {
+    runningBuffer!.startMap(2);
```

Comment:
I think if we were going to do an approach like this, we would want something like a `JsonBufferView` which could encapsulate the logic of the reserved space for the map, with the ability to allocate new objects in the original buffer and overwrite the original pointer values to point at those, so that we could hopefully remove this entire notion of `runningBuffer`.

Yes, it would mean additional allocations but this approach seems too brittle to me.

Reply:
Will experiment :)

I actually like a lot the idea of having access to the `JsonBuffer` that is where we are building up the response to the current request. The idea of "request scope" is heavily used in Guice, it's natural in server or client code to have exactly one of something per request. In Dart I guess we can do that nicely with zones. We can always pass around context explicitly if that turns out to be too brittle.

Separate to that, it occurred to me that we can probably gain a lot by being upfront about the type information that we have.

So for example for `Properties` we know the set of keys and we know the values are all bools.

If we tell `JsonBuffer` what we're doing, it should be possible to write the keys just once, and to pack the bools into bit each. Which will be both smaller and faster :)

Reply:
Also, yes, definitely a split to a type for writing and a type for reading.